#2023EBCS402
syntax = "proto3";

package banking;

import "google/protobuf/timestamp.proto";

service AccountService {
  rpc GetBalance(GetBalanceRequest) returns (GetBalanceResponse);
  rpc UpdateBalance(UpdateBalanceRequest) returns (UpdateBalanceResponse);
}

service TransactionService {
  rpc InitiateTransfer(InitiateTransferRequest) returns (InitiateTransferResponse);
  rpc GetTransactionHistory(GetTransactionHistoryRequest) returns (GetTransactionHistoryResponse);
}

message GetBalanceRequest {
  string user_id = 1;
}

message GetBalanceResponse {
  string user_id = 1;
  double balance = 2;
  string message = 3;
}

message UpdateBalanceRequest {
  string user_id = 1;
  double delta = 2;
  string transaction_id = 3;
}

message UpdateBalanceResponse {
  string user_id = 1;
  double new_balance = 2;
  string message = 3;
}

message InitiateTransferRequest {
  string from_user_id = 1;
  string to_user_id = 2;
  double amount = 3;
}

message InitiateTransferResponse {
  bool success = 1;
  string transaction_id = 2;
  string message = 3;
}

enum TransactionType {
  UNKNOWN = 0;
  TRANSFER = 1;
  CREDIT = 2;
  DEBIT = 3;
}

enum TransactionStatus {
  PENDING = 0;
  COMPLETED = 1;
  FAILED = 2;
}

message Transaction {
  string transaction_id = 1;
  string from_user_id = 2;
  string to_user_id = 3;
  double amount = 4;
  TransactionType type = 5;
  TransactionStatus status = 6;
  google.protobuf.Timestamp created_at = 7;
  string note = 8;
}

message GetTransactionHistoryRequest {
  string user_id = 1;
  int32 limit = 2;
}

message GetTransactionHistoryResponse {
  repeated Transaction transactions = 1;
}
